File: tests/qos/test_service.cpp
// test_service.cpp — Unit tests for medtech::Service abstract interface.
//
// A MockService implements the interface and verifies state transitions:
//   STOPPED → STARTING → RUNNING → (stop) → STOPPING → STOPPED
//   STOPPED → STARTING → FAILED (on simulated error)
// Also verifies stop() is non-blocking while run() is in progress.

#include <atomic>
#include <chrono>
#include <condition_variable>
#include <mutex>
#include <stdexcept>
#include <thread>

#include <gtest/gtest.h>
#include <medtech/service.hpp>

class MockService : public medtech::Service {
public:
    explicit MockService(bool simulate_failure = false)
        : simulate_failure_(simulate_failure)
    {
    }

    void run() override
    {
        state_.store(Orchestration::ServiceState::STARTING);

        if (simulate_failure_) {
            state_.store(Orchestration::ServiceState::FAILED);
            return;
        }

        state_.store(Orchestration::ServiceState::RUNNING);

        // Block until stop() is called
        std::unique_lock lock(mtx_);
        cv_.wait(lock, [this] { return stop_requested_.load(); });

        state_.store(Orchestration::ServiceState::STOPPING);
        // simulate teardown
        state_.store(Orchestration::ServiceState::STOPPED);
    }

    void stop() override
    {
        stop_requested_.store(true);
        cv_.notify_all();
    }

    std::string_view name() const override { return "MockService"; }

    Orchestration::ServiceState state() const override
    {
        return state_.load();
    }

private:
    std::atomic<Orchestration::ServiceState> state_{
        Orchestration::ServiceState::STOPPED};
    std::atomic<bool> stop_requested_{false};
    bool simulate_failure_;
    std::mutex mtx_;
    std::condition_variable cv_;
};

TEST(ServiceInterface, InitialStateIsStopped)
{
    MockService svc;
    EXPECT_EQ(svc.state(), Orchestration::ServiceState::STOPPED);
}

TEST(ServiceInterface, NameIsStable)
{
    MockService svc;
    auto n1 = svc.name();
    auto n2 = svc.name();
    EXPECT_EQ(n1, n2);
    EXPECT_FALSE(n1.empty());
}

TEST(ServiceInterface, FullLifecycleTransitions)
{
    MockService svc;
    EXPECT_EQ(svc.state(), Orchestration::ServiceState::STOPPED);

    // Run in a separate thread
    std::thread runner([&svc] { svc.run(); });

    // Wait for RUNNING state (with timeout)
    auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(5);
    while (svc.state() != Orchestration::ServiceState::RUNNING) {
        ASSERT_LT(std::chrono::steady_clock::now(), deadline)
            << "Timed out waiting for RUNNING state";
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }

    EXPECT_EQ(svc.state(), Orchestration::ServiceState::RUNNING);

    // Stop the service
    svc.stop();
    runner.join();

    EXPECT_EQ(svc.state(), Orchestration::ServiceState::STOPPED);
}

TEST(ServiceInterface, StopIsNonBlocking)
{
    MockService svc;
    std::thread runner([&svc] { svc.run(); });

    // Wait for RUNNING
    auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(5);
    while (svc.state() != Orchestration::ServiceState::RUNNING) {
        ASSERT_LT(std::chrono::steady_clock::now(), deadline);
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }

    // Measure stop() duration — must return within 100ms
    auto start = std::chrono::steady_clock::now();
    svc.stop();
    auto elapsed = std::chrono::steady_clock::now() - start;

    EXPECT_LT(elapsed, std::chrono::milliseconds(100))
        << "stop() blocked for too long";

    runner.join();
}

TEST(ServiceInterface, FailedStateOnError)
{
    MockService svc(/*simulate_failure=*/true);
    EXPECT_EQ(svc.state(), Orchestration::ServiceState::STOPPED);

    svc.run();  // returns after setting FAILED

    EXPECT_EQ(svc.state(), Orchestration::ServiceState::FAILED);
}

File: tests/test_service.py
"""Unit tests for medtech.Service abstract interface.

A MockService implements the interface and verifies state transitions:
  STOPPED -> STARTING -> RUNNING -> (stop) -> STOPPING -> STOPPED
  STOPPED -> STARTING -> FAILED (on simulated error)
Also verifies stop() is non-blocking while run() is in progress.

Tags: @unit @orchestration
"""

from __future__ import annotations

import asyncio
import time

import pytest
from medtech.service import Service, ServiceState


class MockService(Service):
    """Concrete mock implementing medtech.Service for testing."""

    def __init__(self, *, simulate_failure: bool = False) -> None:
        self._state = ServiceState.STOPPED
        self._stop_event: asyncio.Event | None = None
        self._simulate_failure = simulate_failure

    async def run(self) -> None:
        self._state = ServiceState.STARTING

        if self._simulate_failure:
            self._state = ServiceState.FAILED
            return

        self._stop_event = asyncio.Event()
        self._state = ServiceState.RUNNING
        await self._stop_event.wait()

        self._state = ServiceState.STOPPING
        # simulate teardown
        self._state = ServiceState.STOPPED

    def stop(self) -> None:
        if self._stop_event is not None:
            self._stop_event.set()

    @property
    def name(self) -> str:
        return "MockService"

    @property
    def state(self) -> ServiceState:
        return self._state


class TestServiceInitialState:
    def test_initial_state_is_stopped(self):
        svc = MockService()
        assert svc.state == ServiceState.STOPPED

    def test_name_is_stable(self):
        svc = MockService()
        assert svc.name == svc.name
        assert len(svc.name) > 0


class TestServiceLifecycle:
    @pytest.mark.asyncio
    async def test_full_lifecycle_transitions(self):
        svc = MockService()
        assert svc.state == ServiceState.STOPPED

        task = asyncio.create_task(svc.run())

        # Yield to let run() progress
        await asyncio.sleep(0.01)
        assert svc.state == ServiceState.RUNNING

        svc.stop()
        await task

        assert svc.state == ServiceState.STOPPED

    @pytest.mark.asyncio
    async def test_stop_is_non_blocking(self):
        svc = MockService()
        task = asyncio.create_task(svc.run())

        await asyncio.sleep(0.01)
        assert svc.state == ServiceState.RUNNING

        start = time.monotonic()
        svc.stop()
        elapsed = time.monotonic() - start

        assert elapsed < 0.1, f"stop() blocked for {elapsed:.3f}s"

        await task

    @pytest.mark.asyncio
    async def test_failed_state_on_error(self):
        svc = MockService(simulate_failure=True)
        assert svc.state == ServiceState.STOPPED

        await svc.run()

        assert svc.state == ServiceState.FAILED


class TestServiceStateIsIDLGenerated:
    def test_state_returns_idl_type(self):
        svc = MockService()
        assert isinstance(svc.state, ServiceState)

    def test_all_enum_values_exist(self):
        assert hasattr(ServiceState, "STOPPED")
        assert hasattr(ServiceState, "STARTING")
        assert hasattr(ServiceState, "RUNNING")
        assert hasattr(ServiceState, "STOPPING")
        assert hasattr(ServiceState, "FAILED")
        assert hasattr(ServiceState, "UNKNOWN")
